_posts/2020-05-03-flink-sql-demo-building-e2e-streaming-application.md

```diff
@@ -0,0 +1,338 @@
+    item_id BIGINT,
+    category_id BIGINT,
+    behavior STRING,
+    ts TIMESTAMP(3),
+    proctime AS PROCTIME(),   -- generates processing-time attribute using computed column
+    WATERMARK FOR ts AS ts - INTERVAL '5' SECOND  -- defines watermark on ts column, marks ts as event-time attribute
+) WITH (
+    'connector.type' = 'kafka',  -- using kafka connector
+    'connector.version' = 'universal',  -- kafka version, universal supports Kafka 0.11+
+    'connector.topic' = 'user_behavior',  -- kafka topic
+    'connector.startup-mode' = 'earliest-offset',  -- reading from the beginning
+    'connector.properties.zookeeper.connect' = 'localhost:2181',  -- zookeeper address
+    'connector.properties.bootstrap.servers' = 'localhost:9092',  -- kafka broker address
+    'format.type' = 'json'  -- the data format is json
+);
+```
+
+The above snippet declares five fields based on the data format. In addition, it uses the computed column syntax and built-in `PROCTIME()` function to declare a virtual column that generates the processing-time attribute. It also uses the WATERMARK syntax to declare the watermark strategy on the `ts` field (tolerate 5-seconds out-of-order). Therefore, the `ts` field becomes an event-time attribute. For more information about time attributes and DDL syntax, see the following official documents:
+
+- [For time attributes](https://ci.apache.org/projects/flink/flink-docs-release-1.10/dev/table/streaming/time_attributes.html)
+- [For SQL DDL](https://ci.apache.org/projects/flink/flink-docs-release-1.10/dev/table/sql/create.html#create-table)
+
+After creating the `user_behavior` table in the SQL CLI, run `show tables;` and `describe user_behavior;` to see registered tables and table details. Also, run the command `SELECT * FROM user_behavior;` directly in the SQL CLI to preview the data (press `q` to exit).
+
+Next, we discover more about Flink SQL through three real scenarios.
+
+# Hourly Trading Volume
+
+## Create Elasticsearch table using DDL
+
+Let’s create an Elasticsearch result table in the SQL CLI. We need two columns in this case: `hour_of_day` and `buy_cnt` (trading volume).
+
+```sql
+CREATE TABLE buy_cnt_per_hour (
+    hour_of_day BIGINT,
+    buy_cnt BIGINT
+) WITH (
+    'connector.type' = 'elasticsearch', -- using elasticsearch connector
+    'connector.version' = '6',  -- elasticsearch version, 6 supports both 6+ and 7+
+    'connector.hosts' = 'http://localhost:9200',  -- elasticsearch address
+    'connector.index' = 'buy_cnt_per_hour',  -- elasticsearch index name, similar to database table name
+    'connector.document-type' = 'user_behavior', -- elasticsearch type name
+    'connector.bulk-flush.max-actions' = '1',  -- refresh for every row
+    'format.type' = 'json',  -- output data in json format
+    'update-mode' = 'append'
+);
+```
+
+There is no need to create the `buy_cnt_per_hour` index in Elasticsearch in advance since Elasticsearch will automatically create the index if it doesn’t exist.
```

Comment:
```suggestion
There is no need to create the `buy_cnt_per_hour` index in Elasticsearch, in advance since Elasticsearch will automatically create the index if it does not exist.
```

Reply:
Do we need to add `,` before `in advance`? In my understanding, it belongs to the "create index".

Reply:
you're correct @wuchong - this is not necessary here. I will update this suggestion